src/chef_authn.erl

```diff
@@ -446,7 +452,39 @@ do_authenticate_user_request(GetHeader, Method, Path, Body, PublicKey, TimeSkew)
     BodyHash = hashed_body(Body),
     Plain = canonicalize_request(BodyHash, UserId, Method, ReqTime,
                                  Path, SignAlgorithm, SignVersion),
-    verify_sig(Plain, BodyHash, ContentHash, AuthSig, UserId, PublicKey, SignVersion).
+    verify_sig_or_sigs(Plain, BodyHash, ContentHash, AuthSig, UserId, PublicKey, SignVersion).
+
+
+-spec verify_sig_or_sigs(binary(), binary(), binary(), binary(), binary(),
+                         [{key_desc(), public_key_data() | public_key:rsa_public_key()}], binary()) ->
+                                {name, user_id()} | {name, user_id(), key_desc()}.
+verify_sig_or_sigs(Plain, BodyHash, ContentHash, AuthSig, UserId,  [Head | _] = KeyData, SignVersion)
+  when is_tuple(Head) ->
+    verify_sigs(Plain, BodyHash, ContentHash, AuthSig, UserId, KeyData, SignVersion);
+verify_sig_or_sigs(Plain, BodyHash, ContentHash, AuthSig, UserId, KeyData, SignVersion) ->
+    verify_sig(Plain, BodyHash, ContentHash, AuthSig, UserId, KeyData, SignVersion).
+
+
+%% @doc Try sigs in order, returning key_id for the first one that validates correctly
+%% The key_id is an identifier meaningful to the caller. (In our case it might be a tuple {name, userorclient, id, authz_id}
+%% There are some optimization opportunities around hoisting more common stuff out of verify_sig to minimize
+%% what we are repeating each step
+%%
+-spec verify_sigs(binary(), binary(), binary(), binary(), binary(),
+                  [{key_desc(), public_key_data() | public_key:rsa_public_key()}], binary()) ->
+                         {name, user_id(), key_desc()}.
+
+verify_sigs(Plain, BodyHash, ContentHash, AuthSig, UserId, [{KeyId, PubKey}], SignVersion) ->
+    {name, UserId} = verify_sig(Plain, BodyHash, ContentHash, AuthSig, UserId, PubKey, SignVersion),
+    {name, UserId, KeyId};
+verify_sigs(Plain, BodyHash, ContentHash, AuthSig, UserId, [{KeyId, PubKey} | Remaining], SignVersion) ->
+    try
+        {name, UserId} = verify_sig(Plain, BodyHash, ContentHash, AuthSig, UserId, PubKey, SignVersion),
+        {name, UserId, KeyId}
+    catch
+        %% We can fail for other reasons than signature not being right; most of those aren't badmatch though.
+        error:{badmatch, _} -> verify_sigs(Plain, BodyHash, ContentHash, AuthSig, UserId, Remaining, SignVersion)
+    end.
```

Comment:
I'd almost rather push this try / catch logic down into `verify_sig` and have that method return something actually meaningful instead of `badmatch` when it couldn't verify.